src/Grpc.AspNetCore.Server/Internal/PipeExtensions.cs

```diff
@@ -55,6 +55,24 @@ public static Task WriteMessageAsync(this PipeWriter pipeWriter, byte[] messageD
                 return Task.FromException(new RpcException(SendingMessageExceedsLimitStatus));
             }
 
+            // Must call StartAsync before the first pipeWriter.GetSpan() in WriteHeader
+            var startAsyncTask = serverCallContext.HttpContext.Response.StartAsync();
```

Comment:
`WriteMessageAsync` can be called multiple times for a request when streaming. I would prefer that we explicitly check for `Response.HasStarted` before calling `Response.StartAsync()`. First, I think calling StartAsync on the response multiple times feels wrong. Second, it would slightly improve performance by not calling into the feature and having it no-op, and then checking the task.

Reply:
Also when https://github.com/aspnet/AspNetCore/issues/7779 goes into Kestrel then we should only attempt to call `StartAsync` if we're not buffering.

If we're buffering then not calling `StartAsync` would allow the method to modify the response headers with `WriteResponseHeadersAsync` after calling `WriteMessageAsync`.

Reply:
To clarify, when you are buffering, are you replacing Kestrel's PipeWriter with a buffering one?